File: src/com/gmail/marcosav2010/config/Configuration.java
package com.gmail.marcosav2010.config;

import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Properties;

import com.gmail.marcosav2010.logger.ILog;
import com.gmail.marcosav2010.logger.Log;
import com.gmail.marcosav2010.logger.Logger;

public class Configuration implements IConfiguration {

	private final ILog log;

	private String configName;
	private Path path;

	private boolean save;

	private Properties properties;

	public Configuration(String configName) {
		this.configName = configName + ".properties";
		save = false;
		log = new Log(Logger.getGlobal(), "Conf-" + configName);
	}

	public void load(Properties defaultProperties) {
		properties = defaultProperties;

		File f = new File(configName);
		path = f.toPath();
		if (f.exists()) {

			try {
				properties.load(Files.newInputStream(path));
			} catch (IOException e) {
				log.log(e);
			}

		} else {
			try {
				f.createNewFile();
				_store();
			} catch (IOException e) {
				log.log(e);
			}
		}
	}

	@Override
	public boolean exists(String key) {
		return properties.getProperty(key) != null;
	}

	@Override
	public String get(String key) {
		return properties.getProperty(key);
	}

	@Override
	public String get(String key, String def) {
		return properties.getProperty(key, def);
	}

	@Override
	public void set(String key, String value) {
		properties.setProperty(key, value);
		save = true;
	}

	public void store() throws IOException {
		if (save)
			_store();
	}

	private void _store() throws IOException {
		try (OutputStream out = Files.newOutputStream(path)) {
			properties.store(out, null);
		}
	}
}

File: cli/src/main/java/com/gmail/marcosav2010/config/Configuration.kt
package com.gmail.marcosav2010.config

import com.gmail.marcosav2010.logger.ILog
import com.gmail.marcosav2010.logger.Log
import com.gmail.marcosav2010.logger.Logger
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.util.Properties

open class Configuration(configName: String) : IConfiguration {

    private val log: ILog = Log(Logger.global, "Conf-$configName")
    private val configName = "$configName.properties"
    private lateinit var path: Path
    private var save: Boolean = false
    private lateinit var properties: Properties

    fun load(defaultProperties: Properties) {
        properties = defaultProperties
        val f = File(configName)
        path = f.toPath()

        if (f.exists()) {
            try {
                properties.load(Files.newInputStream(path))
            } catch (e: IOException) {
                log.log(e)
            }
        } else {
            try {
                f.createNewFile()
                internalStore()
            } catch (e: IOException) {
                log.log(e)
            }
        }
    }

    override fun exists(key: String) = properties.getProperty(key) != null

    override fun get(key: String): String? = properties.getProperty(key)

    override fun get(key: String, def: String?): String? = properties.getProperty(key, def)

    override fun set(key: String, value: String?) = properties.setProperty(key, value).let { save = true }

    fun store() {
        if (save) internalStore()
    }

    private fun internalStore() = Files.newOutputStream(path).use { out -> properties.store(out, null) }
}